examples/motoko_examples/basic-dao/src/basic_dao.ts

```diff
@@ -0,0 +1,334 @@
+    }
+    if (account) {
+      if (proposal.voters.some((e) => e === caller)) {
+        return yield { err: 'Already voted' };
+      }
+
+      let votes_yes = proposal.votes_yes.amount_e8s;
+      let votes_no = proposal.votes_no.amount_e8s;
+      switch (args.vote) {
+        case Types.Vote.yes: {
+          votes_yes += account.amount_e8s;
+        }
+        case Types.Vote.no: {
+          votes_no += account.amount_e8s;
+        }
+      }
+      proposal.voters.push(caller);
+      let voters = proposal.voters;
+
+      if (votes_yes >= system_params.proposal_vote_threshold.amount_e8s) {
+        // Refund the proposal deposit when the proposal is accepted
+        let account = account_get(proposal.proposer)!;
+        let refunded =
+          account.amount_e8s +
+          system_params.proposal_submission_deposit.amount_e8s;
+        account_put(proposal.proposer, { amount_e8s: refunded });
+
+        state = Types.ProposalState.accepted;
+      }
+
+      if (votes_no >= system_params.proposal_vote_threshold.amount_e8s) {
+        state = Types.ProposalState.rejected;
+      }
+
+      let updated_proposal = {
+        id: proposal.id,
+        votes_yes: { amount_e8s: votes_yes },
+        votes_no: { amount_e8s: votes_no },
+        voters,
+        state,
+        timestamp: proposal.timestamp,
+        proposer: proposal.proposer,
+        payload: proposal.payload,
+      };
+      proposal_put(args.proposal_id, updated_proposal);
+    }
+
+    return yield { ok: state };
+  }
+}
+
+/// Get the current system params
+export function* get_system_params(): Generator<Query<Types.SystemParams>> {
+  return yield system_params;
+}
+
+/// Update system params
+///
+/// Only callable via proposal execution
+export function* update_system_params(
+  payload: Types.UpdateSystemParamsPayload
+): UpdateAsync<void> {
+  let caller = ic.caller();
+  //is  ic.id() === Principal.fromActor(Self) ?
+  if (caller != ic.id()) {
+    return;
+  }
+  system_params = {
+    ...system_params,
+    ...payload,
+  };
+}
+
+/// Deduct the proposal submission deposit from the caller's account
+function deduct_proposal_submission_deposit(
+  caller: Principal
+): Result<void, string> {
+  let from_tokens = account_get(caller);
+
+  if (from_tokens === null) {
+    return { err: 'Caller needs an account to submit a proposal' };
+  }
+  if (from_tokens) {
+    let threshold = system_params.proposal_submission_deposit.amount_e8s;
+    if (from_tokens.amount_e8s < threshold) {
+      return {
+        err: `Caller's account must have at least ${threshold} to submit a proposal`,
+      };
+    } else {
+      let from_amount: int32 /* TODO nat */ =
+        from_tokens.amount_e8s - threshold;
+      account_put(caller, { amount_e8s: from_amount });
+      return { ok: void 0 };
+    }
+  }
+
+  return { err: 'Unexpected error' };
+}
+
+/// Execute all accepted proposals
+function* execute_accepted_proposals(): Generator<
```

Comment:
This function will be the only one that needs to return a generator, along with execute_proposal. But cross canister calls are not coming until beta (which is extremely imminent).

Async should become much clearer with the beta and the new examples that I'll release. There will also be a lot of documentation coming.